### ISSUE TITLE
Idea: Add an type inferer that children can be typed

### ISSUE THREAD

**Issue by iduuck** [FormidableLabs/groqd#6]
Imagine the following scenario:

```tsx
const Article = (props: /* ??? */) => {
	// ... Component
};

export default function ArticlePage({ articles }: InferGetStaticPropsType<typeof getStaticProps>) {
	// ... Component
}

export async function getStaticProps() {
	const articles = await runQuery(/* Some query */);

	return { props: { articles } };
}
```

How to **properly** type the `Article` component? An idea would be to have an `InferType` function – like the underlying zod – so that we are able to also type components, based on the schema, that was defined.

Like this:

```tsx
const articleSchema = q.grab({
	title: q.string(),
})

const Article = (props: q.InferType<typeof articleSchema>) => {
    //                  ^ This would be awesome!
	// ... Component
};

export default function ArticlePage({ articles }: InferGetStaticPropsType<typeof getStaticProps>) {
	// ... Component
}

export async function getStaticProps() {
	const articles = await runQuery(
		q("*", q.filter(/* whatever */), articleSchema),
	);

	return { props: { articles } };
}
```

Sure, I **could** use the inferer of the zod package, but for this, I would need to install zod into my packages, so that typescript is properly picking this up.

**Comment by gksander** [comment]
This is a great idea! I like the idea of shipping a type utility like Zod does, as I see this being a pretty common scenario. I'll work on drafting something up for this.

A quick aside: just this AM I cut a `0.0.5` release, which exports the `z` object from Zod, since Zod is a dependency for this library. This allows you to do `import { z } from "groqd"` without having to install Zod manually.

**Comment by gksander** [comment]
@iduuck I just filed a PR #7 to export such a utility, something that would look something like this:

```ts
const query = q("*", q.grab({ title: q.string() }));
type Article = InferType<typeof query>[number]; // -> { title: string }
```

I _think_ that would do the trick for this case of running a query on Next.js data fetching and wanting the exact type for the React component that'll use the page data. What do you think?

**Comment by gksander** [comment]
Closed via #7
